Replace Zend_Mail (ZF1) with Zend\Mail (ZF2)

schmengler opened the pull request:
This PR probably needs more work, at least more testing, I cannot believe that it is so easy.

Opening for review and acceptance tests. If it's working I plan to refactor the code a bit (see TODO comments) - feedback appreciated!

schmengler committed:
Install ZF 2 mail component

schmengler committed:
Replace ZF1 Zend_Mail with ZF2 Zend\Mail

schmengler committed:
Remove unused variable

schmengler committed:
Update unit tests for ZF 2 mail

orlangur commented:
> needs more work, at least more testing, I cannot believe that it is so easy

True, true... One integration test is failing, another one caused fatal error :) Check [Travis CI](https://travis-ci.org/magento/magento2/builds/203273924).

Note that refactoring should be performed in such a way that all existing customizations relying on `Magento_Framework` and not `Zend_Mail` directly work as before (just like integration tests, people will not be happy getting such errors after upgrade).

Will share my thoughts on the code a bit later.

schmengler commented:
> True, true... One integration test is failing, another one caused fatal error :) Check Travis CI.

I've seen it, but could not look into the details yet. I tried to take care not to change the `Magento_Framework` behavior, and it's possible that tests are only failing because mocks have to be changed.

I'm waiting for your feedback on the code now!

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -7,14 +7,25 @@
+use Zend\Mime\Part;
+
+/**
+ * @todo composition instead of inheritance for better testability
```
Do it now and expose only necessary public methods.

Otherwise people may rely on Zend implementation rather then on narrowed interface and we will continue to have problems like Fatal Error with absent  `getBodyHtml`.

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -7,14 +7,25 @@
     public function __construct($charset = 'utf-8')
     {
-        parent::__construct($charset);
+        $this->encoding = $charset;
```
Rename constructor argument accordingly and use `setEncoding` method of `\Zend\Mail\Message` instead (currently `$this->getHeaders()->setEncoding($encoding);` is missed).

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -25,35 +36,40 @@ public function __construct($charset = 'utf-8')
      * @return $this
      */
-    public function setBody($body)
+    public function setMessageType($type)
```
Please preserve order of methods to simplify checking of correspondence during review (getBody -> getBody, setBody -> setBody etc).

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -25,35 +36,40 @@ public function __construct($charset = 'utf-8')
     {
-        return $this->messageType == self::TYPE_TEXT ? $this->getBodyText() : $this->getBodyHtml();
+        if (is_string($body) && $this->messageType === MessageInterface::TYPE_HTML) {
+            $body = self::htmlMimeFromString($body);
```
Method name must begin with verb, I suggest `createHtmlMimeFromString`.

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -25,35 +36,40 @@ public function __construct($charset = 'utf-8')
-    public function getBody()
+    public function setBody($body)
     {
-        return $this->messageType == self::TYPE_TEXT ? $this->getBodyText() : $this->getBodyHtml();
```
Calls of `getBodyHtml` must be fixed all over the code, this is why we saw Fatal Error on Travis. Integration tests do not use mocks usually.

schmengler replied:
Fortunately `getBodyHtml` is only used in integration tests. I'm trying to fix them

orlangur commented on lib/internal/Magento/Framework/Mail/Test/Unit/MessageTest.php:
```diff
@@ -14,6 +14,7 @@ class MessageTest extends \PHPUnit_Framework_TestCase
 
     protected function setUp()
     {
+        $this->markTestSkipped('obsolete ZF 1 test');
```
Test must be reworked within this pull request.

orlangur commented on lib/internal/Magento/Framework/Mail/Test/Unit/Template/TransportBuilderTest.php:
```diff
@@ -53,6 +54,7 @@ protected function setUp()
         $objectManagerHelper = new \Magento\Framework\TestFramework\Unit\Helper\ObjectManager($this);
         $this->templateFactoryMock = $this->getMock(\Magento\Framework\Mail\Template\FactoryInterface::class);
         $this->messageMock = $this->getMock(\Magento\Framework\Mail\Message::class);
+        $this->messageMock->method('getHeaders')->willReturn(new Headers());
```
`Headers` class must be mocked as well.

orlangur commented on lib/internal/Magento/Framework/Mail/Test/Unit/TransportTest.php:
```diff
@@ -5,45 +5,32 @@
-            ->method('getParts')
-            ->will($this->returnValue(['a', 'b']));
+        $transport = new \Magento\Framework\Mail\Transport(
+            new \Magento\Framework\Mail\Message()
```
`Message` class must be mocked here.

schmengler replied:
I disagree. It's part of the same component and not using external resources. To test handling of an incomplete message, we create a real message without "to" header. A stub just complicates things and a mock will make unnecessary assumptions about the implementation details.

orlangur commented on lib/internal/Magento/Framework/Mail/Transport.php:
```diff
@@ -6,7 +6,8 @@
 namespace Magento\Framework\Mail;
 
-class Transport extends \Zend_Mail_Transport_Sendmail implements \Magento\Framework\Mail\TransportInterface
+
```
Useless new line and `phpcs` does not enforce it yet.

orlangur commented on lib/internal/Magento/Framework/Mail/Transport.php:
```diff
@@ -6,7 +6,8 @@
 
-class Transport extends \Zend_Mail_Transport_Sendmail implements \Magento\Framework\Mail\TransportInterface
+
+class Transport extends \Zend\Mail\Transport\Sendmail implements \Magento\Framework\Mail\TransportInterface
```
Do aggregation instead of inheritance.

orlangur commented on lib/internal/Magento/Framework/Mail/Transport.php:
```diff
@@ -20,8 +21,8 @@ class Transport extends \Zend_Mail_Transport_Sendmail implements \Magento\Framew
     {
-        if (!$message instanceof \Zend_Mail) {
-            throw new \InvalidArgumentException('The message should be an instance of \Zend_Mail');
+        if (!$message instanceof \Zend\Mail\Message) {
```
I don't think we should keep classes inherited from `Zend` classes in the system.

Just create necessary instance of `\Zend\Mail\Message` in `sendMessage` method instead.

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -7,14 +7,25 @@
+
+/**
+ * @todo composition instead of inheritance for better testability
+ * - add a ZendMailDecorator interface with getZendMail() method for usage in \Magento\Framework\Mail\Transport
```
Please no. To me it looks like `\Zend\Mail\Message` has all necessary setters already.

orlangur commented on lib/internal/Magento/Framework/Mail/Message.php:
```diff
@@ -7,14 +7,25 @@
+ * @todo composition instead of inheritance for better testability
+ * - add a ZendMailDecorator interface with getZendMail() method for usage in \Magento\Framework\Mail\Transport
+ * @todo get rid of temporal coupling (setMessageType() + setBody())
+ * - deprecate setMessageType(), implement a HtmlMessage decorator instead
```
I don't think such decorator would be really valuable. Do we have some needed cases not covered with current text/html types implementation?

schmengler replied:
The problem with the current implementation is that you have to call `setMessageType()` before `setBody()`. I'm now considering changing the `MessageInterface` interface. As it should be used via `TransportBuilder`, it should not have serious impact.

- The current signatures `setBody(mixed $body)` and `mixed getBody()` are not meaningful. Either "mixed" must be replaced with "string" or we are leaking the zend abstraction.
- The interface does not allow multitype mails (this is the needed case not covered. Something that Magento always has been lacking of)

One idea is, replacing `setMessageType()`, `setBody()` and `getBody()` with:

- `setBodyHtml(string $html)`
- `setBodyText(string $text)`
- `string getBodyText()`
- `string getBodyHtml()`

This allows text mails, HTML mails, and multitype mails, does not require calling them in any specific order and does not reveal anything of the underlying Zend mail component.

schmengler replied:
Another advantage is that this resembles the old zf1 implementation, so 3rd party code is actually less likely to break even though we have to change an `@api` interface

orlangur commented:
@schmengler, thanks for your efforts, implementation in general looks quite clean and minimalistic (just does what is really necessary).

Please make sure all builds are green before next review iteration.

okorshenko commented:
Hi @schmengler 
Are you going to proceed this with PR?

schmengler commented:
@okorshenko yes as soon as I find some time for it

okorshenko commented:
@schmengler thank you. Please let us know when you  will be ready. Thank you

schmengler committed:
Merge branch 'develop' into zend-mail-2

Conflicts:
	composer.lock

schmengler committed:
Remove zend mail inheritance

schmengler commented:
Because it's hidden now in the "outdated" review, here again my concerns about the message interface:

The problem with the current implementation is that you have to call `setMessageType()` before `setBody()`. I'm now considering changing the `MessageInterface` interface. As it should be used via `TransportBuilder`, it should not have serious impact.

- The current signatures `setBody(mixed $body)` and `mixed getBody()` are not meaningful. Either "mixed" must be replaced with "string" or we are leaking the zend abstraction.
- The interface does not allow multitype mails (this is the needed case not covered. Something that Magento always has been lacking of)

One idea is, replacing `setMessageType()`, `setBody()` and `getBody()` with:

- `setBodyHtml(string $html)`
- `setBodyText(string $text)`
- `string getBodyText()`
- `string getBodyHtml()`

This allows text mails, HTML mails, and multitype mails, does not require calling them in any specific order and does not reveal anything of the underlying Zend mail component.

Another advantage is that this resembles the old zf1 implementation, so 3rd party code is actually less likely to break even though we have to change an `@api` interface

I already added one method `getRawMessage()` because otherwise it would not be possible to have a Zend transport implementation by only using the message interface and not the underlying Zend message, so the PR will cause a BC break anyways. But IMHO it's a minor one and it is necessary to get the mail interfaces right.

schmengler commented:
And while we are at it, shouldn't `TransportInterface` look like this:

```
    function sendMessage(MessageInterface $message);
```

instead of 

```
    function sendMessage()
```

?

I don't understand, why the "Transport" implementation receives the message in the constructor. I'd rather have the transport instance reusable for different messages.

With the current implementation we have the `TransportBuilder` class, which actually builds a message. That would be much simpler and clearer if we did not couple the transport to the message like currently.

schmengler committed:
Remove conflicting class imports

schmengler committed:
Fix integration test and code style issues

schmengler committed:
Remove redundant unit test

The transport now can take any implementation of the message interface as argument

Commit:
Merge branch 'develop' into zend-mail-2

Commit:
Merge branch 'develop' into zend-mail-2

Commit:
Merge branch 'develop' into zend-mail-2

Commit:
MAGETWO-67680: Replace Zend_Mail (ZF1) with Zend\Mail (ZF2) #8608

 - updated composer.lock file

okorshenko commented:
Hi @schmengler 
Just FYI: we stabilized this branch and working on the testing of new changes.

0m3r commented:
What about getHeaders and setHeaders for \Magento\Mail\Message? 

I want to set headers like a in-reply-to, message-id, references etc.

klein0r commented:
Is there a reason why this has been included into 2.2.8?

orlangur commented:
@klein0r yep, backward compatibility is the reason.
